File: migrations/devnet/initializeProgram.ts
// Migrations are an early feature. Currently, they're nothing more than this
// single deploy script that's invoked from the CLI, injecting a provider
// configured from the workspace's Anchor.toml.

import {
  BN,
  Program,
  Provider,
  setProvider,
  Wallet,
} from "@project-serum/anchor";
import {
  Connection,
  ConfirmOptions,
  Keypair,
  SystemProgram,
  PublicKey,
  clusterApiUrl,
  LAMPORTS_PER_SOL,
} from "@solana/web3.js";

// Initialize Web3 devnet provider.
const idl = require("../../target/idl/faktor.json");
const opts: ConfirmOptions = {
  preflightCommitment: "processed",
};
const wallet = new Wallet(Keypair.generate());
const endpoint = process.env.LOCALNET
  ? "http://localhost:8899"
  : clusterApiUrl("devnet");
const connection = new Connection(endpoint, opts.preflightCommitment);
const provider = new Provider(connection, wallet, opts);
const programID = new PublicKey(idl.metadata.address);
const program = new Program(idl as any, programID, provider);
setProvider(provider);

// Seeds
const SEED_CONFIG = Buffer.from("config");
const SEED_PROCESSOR = Buffer.from("processor");
const SEED_STACK = Buffer.from("stack");
const SEED_TREASURY = Buffer.from("treasury");

// Initial data.
const ONE_MINUTE = 60;
let now = new Date(new Date().setSeconds(0, 0));
const time1 = dateToSeconds(now);
const time2 = time1 + ONE_MINUTE;
const time3 = time2 + ONE_MINUTE;

// PDA type
type PDA = {
  address: PublicKey;
  bump: number;
};

// PDAs
let configPDA: PDA,
  processorPDA: PDA,
  treasuryPDA: PDA,
  stack1PDA: PDA,
  stack2PDA: PDA,
  stack3PDA: PDA;

// Helper function to airdrop SOL to an account.
async function airdrop(publicKey: PublicKey, amount: number) {
  console.log(`[INFO] Airdropping ${amount} SOL to ${publicKey.toString()}`);
  await provider.connection
    .requestAirdrop(publicKey, amount * LAMPORTS_PER_SOL)
    .then((sig: any) =>
      provider.connection.confirmTransaction(sig, "confirmed")
    );
}

// Helper function to convert date to UNIX seconds.
export function dateToSeconds(date: Date): number {
  return Math.floor(date.getTime() / 1000);
}

// Helper function to initialize server.
async function initializeAddresses() {
  // Config
  const [configAddress, configBump] = await PublicKey.findProgramAddress(
    [SEED_CONFIG],
    program.programId
  );
  configPDA = {
    address: configAddress,
    bump: configBump,
  };

  // Config
  const [processorAddress, processorBump] = await PublicKey.findProgramAddress(
    [SEED_PROCESSOR],
    program.programId
  );
  processorPDA = {
    address: processorAddress,
    bump: processorBump,
  };

  // Treasury
  const [treasuryAddress, treasuryBump] = await PublicKey.findProgramAddress(
    [SEED_TREASURY],
    program.programId
  );
  treasuryPDA = {
    address: treasuryAddress,
    bump: treasuryBump,
  };

  const [stack1Address, stack1Bump] = await PublicKey.findProgramAddress(
    [SEED_STACK, Buffer.from(`${time1}`)],
    program.programId
  );
  stack1PDA = {
    address: stack1Address,
    bump: stack1Bump,
  };

  // Stack2
  const [stack2Address, stack2Bump] = await PublicKey.findProgramAddress(
    [SEED_STACK, Buffer.from(`${time2}`)],
    program.programId
  );
  stack2PDA = {
    address: stack2Address,
    bump: stack2Bump,
  };

  // Stack3
  const [stack3Address, stack3Bump] = await PublicKey.findProgramAddress(
    [SEED_STACK, Buffer.from(`${time3}`)],
    program.programId
  );
  stack3PDA = {
    address: stack3Address,
    bump: stack3Bump,
  };

  console.log(`[INFO] Config:`, configPDA.address.toString());
  console.log(`[INFO] Processor:`, processorPDA.address.toString());
  console.log(`[INFO] Treasury:`, treasuryPDA.address.toString());
  console.log(`[INFO] Stack 1:`, stack1PDA.address.toString());
  console.log(`[INFO] Stack 2:`, stack2PDA.address.toString());
  console.log(`[INFO] Stack 3:`, stack3PDA.address.toString());
}

// Helper function execute the initialize_program RPC call.
async function initializeProgram() {
  // Config data
  const transferFeeDistributor = 1000;
  const transferFeeProgram = 1000;

  // Execute RPC call
  console.log(`Initializing program...`);
  await program.rpc.initializeProgram(
    new BN(transferFeeDistributor),
    new BN(transferFeeProgram),
    new BN(time1),
    new BN(time2),
    new BN(time3),
    configPDA.bump,
    processorPDA.bump,
    stack1PDA.bump,
    stack2PDA.bump,
    stack3PDA.bump,
    treasuryPDA.bump,
    {
      accounts: {
        signer: wallet.publicKey,
        config: configPDA.address,
        processor: processorPDA.address,
        stackA: stack1PDA.address,
        stackB: stack2PDA.address,
        stackC: stack3PDA.address,
        treasury: treasuryPDA.address,
        systemProgram: SystemProgram.programId,
      },
      signers: [wallet.payer],
    }
  );
}

// Helper function to run the script.
async function run() {
  console.log(
    "[INFO] Initializing Faktor devent program:",
    program.programId.toString()
  );
  await airdrop(wallet.publicKey, 1);
  await initializeAddresses();
  await initializeProgram();
}

run();

File: programs/payment/src/instructions/initialize_program.rs
use {
    anchor_lang::{
        prelude::*,
        solana_program::system_program
    },
    crate::state,
    std::mem::size_of
};

#[derive(Accounts)]
#[instruction(
    config_transfer_fee_distributor: u64,
    config_transfer_fee_program: u64,
    authority_bump: u8,
    config_bump: u8,
    treasury_bump: u8,
)]
pub struct InitializeProgram<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(
        init,
        seeds = [state::SEED_AUTHORITY],
        bump = authority_bump,
        payer = signer,
        space = 8 + size_of::<state::Authority>(),
    )]
    pub authority: Account<'info, state::Authority>,
    
    #[account(
        init, 
        seeds = [state::SEED_CONFIG], 
        bump = config_bump, 
        payer = signer, 
        space = 8 + size_of::<state::Config>()
    )]
    pub config: Account<'info, state::Config>,
    
    #[account(
        init, 
        seeds = [state::SEED_TREASURY], 
        bump = treasury_bump, 
        payer = signer, 
        space = 8 + size_of::<state::Treasury>()
    )]
    pub treasury: Account<'info, state::Treasury>,

    #[account(address = system_program::ID)]
    pub system_program: Program<'info, System>,
}

pub fn handler(
    ctx: Context<InitializeProgram>, 
    config_transfer_fee_distributor: u64,
    config_transfer_fee_program: u64,
    authority_bump: u8,
    config_bump: u8,
    treasury_bump: u8,
) -> ProgramResult {
    // Get accounts.
    let authority = &mut ctx.accounts.authority;
    let config = &mut ctx.accounts.config;
    let treasury = &mut ctx.accounts.treasury;

    // Initialize authority account.
    authority.bump = authority_bump;

    // Initialize config account.
    config.transfer_fee_distributor = config_transfer_fee_distributor;
    config.transfer_fee_program = config_transfer_fee_program;
    config.bump = config_bump;

    // Initialize treasury account.
    treasury.bump = treasury_bump;

    return Ok(());
}
